voidcrew/code/modules/overmap/ships/pvp/subverter.dm

```diff
@@ -0,0 +1,235 @@
+		subvert_bark("Subverter recharging!")
+		return
+	if (world.time < target_ship.engine_cooldown)
+		subvert_bark("Target vessel is already subverted.")
+		return
+	if (target_ship.state == OVERMAP_SHIP_IDLE)
+		subvert_bark("Target vessel is currently docked (outside of hyperspace).")
+		return
+	if (target_ship.state == OVERMAP_SHIP_DOCKING)
+		subvert_bark("Target vessel is currently docking.")
+		return
+	if (target_ship.state == OVERMAP_SHIP_UNDOCKING)
+		subvert_bark("Target vessel is currently undocking.")
+		return
+	if (world.time < target_ship.sub_grace)
+		subvert_bark("Cannot subvert that vessel currently (Grace period).")
+		return
+	if (target_ship.antivirus_nodes > 0)
+		subvert_bark("Agent failed! Reason: Antivirus")
+		return
+	if (!can_subvert(target_ship))
+		subvert_bark("Failed to subvert (Unknown error)")
+		return
+	subvert_bark(bark_success_message, SUBVERTER_BARK_SUCCESS)
+	return TRUE
+
+//quick maths
+/obj/machinery/subverter/proc/can_subvert(obj/structure/overmap/ship/simulated/target_ship, antag = FALSE)
+	return !(((!antag) && (target_ship.prefix == "NEU" || (ship && (ship.prefix == "NEU" || (ship.prefix != "KOS" && ship.prefix == target_ship.prefix))))) || world.time < subverter_cooldown || world.time < target_ship.engine_cooldown || world.time < target_ship.sub_grace || target_ship.state != OVERMAP_SHIP_FLYING)
+
+/obj/machinery/subverter/proc/force_dock(obj/structure/overmap/ship/simulated/target_ship)
+	target_ship.decelerate(target_ship.max_speed)
+	target_ship.dock_in_empty_space(usr)
+	target_ship.stall_engines(sub_engine)
+	target_ship.most_recent_helm.say("CRITICAL ERROR: SYSTEM TAKEOVER")
+	playsound(target_ship.most_recent_helm, 'voidcrew/sound/voice/booterattack.ogg', 100, 0, 0)
+
+/obj/machinery/subverter/proc/attempt_to_subvert(obj/structure/overmap/ship/simulated/target_ship)
+	. = FALSE
+	bark_processing(target_ship)
+	if (can_subvert(target_ship))
+		subverter_cooldown = world.time + sub_recharge
+		its_gettin_hot_in_here()
+		//use_power(active_power_usage)
+		if (!target_ship.run_antivirus())
+			force_dock(target_ship)
+			addtimer(CALLBACK(target_ship, /obj/structure/overmap/ship/simulated/.proc/systems_restored), target_ship.engine_cooldown - world.time)
+			return TRUE
+		else
+			target_ship.sub_blocked(src)
+
+/obj/machinery/subverter/proc/update_stats()
+	sub_recharge = SUBVERTER_RECHARGE_TIME
+	sub_engine = SUBVERTER_ENGINE_STALL_TIME
+	active_power_usage = 1000
+	for(var/obj/item/stock_parts/micro_laser/B in component_parts)
+		sub_engine += (B.rating-1) * (15 SECONDS)
+	for(var/obj/item/stock_parts/manipulator/M in component_parts)
+		sub_recharge -= (M.rating-1) * (5 SECONDS)
+	for(var/obj/item/stock_parts/capacitor/P in component_parts)
+		active_power_usage /= P.rating
+		spiciness /= P.rating
+
+/obj/machinery/subverter/RefreshParts()
+	update_stats()
+
+/obj/machinery/subverter/proc/its_gettin_hot_in_here()
+	//var/turf/T = get_turf(src)
+	var/turf/L = loc
```

Comment:
loc isn't always guaranteed to be a turf, you should definitely use the commented out code above